Sources/AsyncHTTPClient/RequestValidation.swift

```diff
@@ -16,93 +16,79 @@ import NIOCore
         }
 
-        var transferEncoding: String?
-        var contentLength: Int?
-        let encodings = self[canonicalForm: "Transfer-Encoding"].map { $0.lowercased() }
+        try self.validateFieldNames()
+        try Self.validateTransferEncoding(encodings)
 
-        guard !encodings.contains("identity") else {
-            throw HTTPClientError.identityCodingIncorrectlyPresent
+        if contentLength != nil {
+            self.remove(name: "Content-Length")
+        }
+        if !encodings.isEmpty {
+            self.remove(name: "Transfer-Encoding")
         }
 
-        self.remove(name: "Transfer-Encoding")
-
-        try self.validateFieldNames()
+        let connectionClose = self[canonicalForm: "connection"].lazy.map { $0.lowercased() }.contains("close")
 
-        guard let body = body else {
-            self.remove(name: "Content-Length")
+        switch bodyLength {
+        case .fixed(0):
             // if we don't have a body we might not need to send the Content-Length field
             // https://tools.ietf.org/html/rfc7230#section-3.3.2
             switch method {
             case .GET, .HEAD, .DELETE, .CONNECT, .TRACE:
                 // A user agent SHOULD NOT send a Content-Length header field when the request
                 // message does not contain a payload body and the method semantics do not
                 // anticipate such a body.
-                return metadata
+                break
             default:
                 // A user agent SHOULD send a Content-Length in a request message when
                 // no Transfer-Encoding is sent and the request method defines a meaning
                 // for an enclosed payload body.
                 self.add(name: "Content-Length", value: "0")
-                return metadata
             }
-        }
-
-        if case .TRACE = method {
-            // A client MUST NOT send a message body in a TRACE request.
-            // https://tools.ietf.org/html/rfc7230#section-4.3.8
-            throw HTTPClientError.traceRequestWithBody
-        }
-
-        guard (encodings.lazy.filter { $0 == "chunked" }.count <= 1) else {
-            throw HTTPClientError.chunkedSpecifiedMultipleTimes
-        }
-
-        if encodings.isEmpty {
-            if let length = body.length {
-                self.remove(name: "Content-Length")
-                contentLength = length
-            } else if !self.contains(name: "Content-Length") {
-                transferEncoding = "chunked"
+            return .init(connectionClose: connectionClose, body: .fixedSize(0))
+        case .fixed(let length):
+            if case .TRACE = method {
+                // A client MUST NOT send a message body in a TRACE request.
+                // https://tools.ietf.org/html/rfc7230#section-4.3.8
+                throw HTTPClientError.traceRequestWithBody
+            }
+            if encodings.isEmpty {
+                self.add(name: "Content-Length", value: String(length))
```

Comment:
There might be an interesting case here. What if the user handed us a fixed size body payload, but also set the Content-Length header to something different?

Reply:
I once had validation in there but noticed that we actually have tests that check that we always overwrite the `Content-Length` header. I would also prefer to validate it but this might break user code. WDYT?

Reply:
https://github.com/swift-server/async-http-client/blob/62ad8183765c23c48de67aa8ff08b653d679e0b4/Tests/AsyncHTTPClientTests/RequestValidationTests.swift#L43-L49

Reply:
We need @Lukasa's wisdom here.